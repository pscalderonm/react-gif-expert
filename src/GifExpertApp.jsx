import PropTypes from 'prop-types';
import { useState } from "react";
import { AddCategory, GifGrid } from "./components";

export const GifExpertApp = ({defaultCategory=''}) => {
  
  const [categories, setCategories] = useState([defaultCategory]);

  const handleAddCategory = ( newCategory ) =>{
    if (categories.includes(newCategory)){
      return;
    }

    setCategories(( list ) => [ newCategory, ...list ]);
  };

  return (
    <>
      <h1>GitExpertApp</h1>

      <AddCategory
        onNewCategory={ handleAddCategory }
      />

      {/* React no promueve el uso de indices de arrays para manejo de keys */}
      { categories.map((category)=>(
            <GifGrid 
              key={ category } 
              category={ category } 
            />
        )) 
      }
    </>
  )
}


GifExpertApp.propTypes = {
  defaultCategory: PropTypes.string
};