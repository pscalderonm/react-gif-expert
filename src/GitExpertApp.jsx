import { useState } from "react";
import { AddCategory, GifGrid } from "./components";

export const GitExpertApp = () => {
  
  const [categories, setCategories] = useState(['Cowboy Bebop']);

  const handleAddCategory = ( newCategory ) =>{
    if (categories.includes(newCategory)){
      return;
    }

    setCategories(( list ) => [ newCategory, ...list ]);
  };

  return (
    <>
      <h1>GitExpertApp</h1>

      <AddCategory
        onNewCategory={ handleAddCategory }
      />

      {/* React no promueve el uso de indices de arrays para manejo de keys */}
      { categories.map((category)=>(
            <GifGrid 
              key={ category } 
              category={ category } 
            />
        )) 
      }
    </>
  )
}
