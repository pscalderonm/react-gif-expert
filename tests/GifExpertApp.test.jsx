import { render, screen, fireEvent } from "@testing-library/react";
import { GifExpertApp } from "../src/GifExpertApp";

describe('Pruebas en <GifExpertApp/>', () => { 
  test('debe de tener al menos la categoría por defecto renderizada', () => { 

    const defaultCategory = 'Cowboy Bebop';

    render(<GifExpertApp defaultCategory={defaultCategory}/>);

    expect(screen.getByRole('heading', { level: 3 }).innerHTML).toBe(defaultCategory);
   
   });

   test('debe de tener los elementos primarios renderizados', ()=>{

    const {container} = render(<GifExpertApp />);

    expect(container).toMatchSnapshot();
   });

   test('debe de mostrar una única categoría y no duplicados', ()=>{
    const category = 'Cowboy Bebop';

    render(<GifExpertApp defaultCategory={category} />);

    const input = screen.getByRole('textbox');
    const form = screen.getByRole('form');

    fireEvent.input(input, {target:{value: category}});
    fireEvent.submit(form);

    expect(screen.getAllByText(category).length).toBe(1);
   });
 });