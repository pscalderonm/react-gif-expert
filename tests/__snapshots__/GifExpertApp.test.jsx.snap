// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifExpertApp/> debe de tener los elementos primarios renderizados 1`] = `
<div>
  <h1>
    GitExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Search for GIFs"
      type="text"
      value=""
    />
  </form>
  <h3 />
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
